### PR TITLE
[Scala 3] Add infix modifier

### PR DESCRIPTION
**Pull request by tgodzik**
This PR adds the infix modifier as well as makes sure it's properly sorted when using SortModifiers and that it works within extensions.

Infix modifier is mentioned here: https://dotty.epfl.ch/docs/reference/changed-features/operators.html

This also handles https://github.com/scalameta/scalafmt/pull/2298#pullrequestreview-562910763

### CODE REVIEW

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 1]
```diff
@@ -470,8 +470,10 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
-      case tok @ FormatToken(_: T.RightParen, _: T.KwDef | _: T.Comment, _)
-          if leftOwner.is[Defn.ExtensionGroup] =>
+      case tok @ FormatToken(_: T.RightParen, _, _)
+          if leftOwner.is[Defn.ExtensionGroup] && nextNonComment(
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:474]: can you please manually break after && and format rest on one line?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 2]
```diff
@@ -470,8 +470,10 @@ class Router(formatOps: FormatOps) {
...
+            formatToken
+          ).right.isNot[LeftParenOrBrace] =>
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:476]: should this be .is[KwDef]?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 3]
```diff
@@ -470,8 +470,10 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
-      case tok @ FormatToken(_: T.RightParen, _: T.KwDef | _: T.Comment, _)
-          if leftOwner.is[Defn.ExtensionGroup] =>
+      case tok @ FormatToken(_: T.RightParen, _, _)
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: this change, and its tests, seem unrelated to infix, no? should they be part of a different pr?

File: scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat [hunk 4]
```diff
@@ -0,0 +1,12 @@
+maxColumn = 200 #to not disturb the output too much
```
Comment (kitbellew) [inline scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat:1]: do we need a separate .stat file for this test?

**Review by kitbellew — CHANGES_REQUESTED** [review]
i think two unrelated changes are mixed here

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 2 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:476]: So the change is actually related to `infix` as we can get `infix def` etc. So having it as `.isNot[LeftParenOrBrace]` is safer, because we want to match only if we don't have a `{` block or another `(` param list.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:476]: is this cleaner than, say, .is[KwDef] or .is[Mod.xxx]?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 3 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: At least `.isNot[LeftParenOrBrace]` is needed for this change and since the list of accepted tokens started growing it turned out to be easier to just use it together with `nextNonComment`, which simplifies this logic a lot.

Because, this also fixes the issue with the comments I added the tests here. I can add it in a different PR, but it would just be a PR with those tests.

File: scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat (hunk 4 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat:1]: Not really, there were separate files for modifiers already that's why I added another one. I do agree that we can just put it into a single file. I will put it together with `inline`, `transparent` and `open`. What do you think?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 3 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: yes, i understand now. what type do you get for rightOwner of next non comment when it is def, infix, (, {, etc?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 2 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:476]: Yes, because we would need at least 3 with `Ident` as `infix` is a soft keyword. Also, it's more logical `LeftParenOrBrace` means we are not starting a block and not doing another param list, so the only other alternative is that it is the definition.

File: scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat (hunk 4 above)
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat:1]: i like this approach 🙂

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 2 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:476]: thank you, understood. unless there's a more logical way with rightOwner tree, vs right token, I'm ok with this. thanks again for looking into this.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 3 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: it should be `Defn.ExtensionGroup`  for `(`, `Term.Block` for `{` for the others `Defn.Def` - so nothing I think to help us here.

File: scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat (hunk 4 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/resources/rewrite/SortModifiersInfix.stat]: I put them all together into `SortModifiersScala3.stat`

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 1 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Fixed!

**Review by kitbellew — APPROVED** [review]
